File: Kotlin/problem012.kt
// Problem 12 - Project Euler
// http://projecteuler.net/index.php?section=problems&id=12
import kotlin.math.*

fun count_divisors(x: Int): Int {
    val lim = ceil(sqrt(x.toDouble())).toInt()
    return (1..lim)
            .map { n ->
                if (x % n == 0) listOf(n, x/n) else null
            }
            .filterNotNull()
            .flatten()
            .distinct()
            .count()
}

fun Int.triangle() = this * (this + 1) / 2

fun main(args: Array<String>) {
    var n = 1
    while (count_divisors(n.triangle()) <= 500) {
        n++
    }
    println("${n.triangle()}")
}

File: TypeScript/problem012.ts
/*
 * #12 Highly Divisible Triangular Number - Project Euler
 * http://projecteuler.net/problem=12
 */
{
    function countDivisors(n: number): number {
        let count = 0;

        for (let i = 1; i * i <= n; i++) {
            if (n % i == 0) {
                count += (n / i != i) ? 2 : 1;
            }
        }

        return count;
    }

    let num = 1;
    let i = 1;
    while (true) {
        const count = countDivisors(num);
        if (count > 500) {
            console.log(num, count);
            break;
        }
        i += 1;
        num += i;
    }
}
